nym-vpn-core/nym-vpn-lib/src/mobile/runner.rs

```diff
@@ -0,0 +1,427 @@
+    task_manager: TaskManager,
+    generic_config: GenericNymVpnConfig,
+    #[cfg(target_os = "ios")]
+    tun_provider: Arc<dyn OSTunProvider>,
+    shutdown_token: CancellationToken,
+}
+
+impl WgTunnelRunner {
+    pub fn new(config: VPNConfig, shutdown_token: CancellationToken) -> Result<Self> {
+        let user_agent = UserAgent::from(nym_bin_common::bin_info_local_vergen!());
+        tracing::info!("User agent: {user_agent}");
+
+        let generic_config = GenericNymVpnConfig {
+            mixnet_client_config: MixnetClientConfig {
+                enable_poisson_rate: false,
+                disable_background_cover_traffic: false,
+                enable_credentials_mode: false,
+                min_mixnode_performance: None,
+                min_gateway_performance: None,
+            },
+            data_path: config.credential_data_path,
+            gateway_config: nym_gateway_directory::Config {
+                api_url: config.api_url,
+                nym_vpn_api_url: config.vpn_api_url,
+                min_gateway_performance: None,
+            },
+            entry_point: EntryPoint::from(config.entry_gateway),
+            exit_point: ExitPoint::from(config.exit_router),
+            nym_ips: None,
+            nym_mtu: None,
+            dns: None,
+            disable_routing: false,
+            user_agent: Some(user_agent.clone()),
+        };
+
+        let task_manager = TaskManager::new(TASK_MANAGER_SHUTDOWN_TIMER_SECS).named("nym_vpn_lib");
+        let gateway_directory_client =
+            GatewayClient::new(generic_config.gateway_config.clone(), user_agent.clone()).map_err(
+                |err| GatewayDirectoryError::FailedtoSetupGatewayDirectoryClient {
+                    config: Box::new(generic_config.gateway_config.clone()),
+                    source: err,
+                },
+            )?;
+
+        Ok(Self {
+            gateway_directory_client,
+            task_manager,
+            generic_config,
+            #[cfg(target_os = "ios")]
+            tun_provider: config.tun_provider,
+            shutdown_token,
+        })
+    }
+
+    pub async fn start(mut self) -> Result<()> {
+        let SelectedGateways { entry, exit } = self.select_gateways().await?;
+        let mixnet_client = self.start_mixnet_client(&entry).await?;
+
+        let auth_addresses = match self.setup_auth_addresses(&entry, &exit) {
+            Ok(auth_addr) => auth_addr,
+            Err(err) => {
+                // Put in some manual error handling, the correct long-term solution is that handling
+                // errors and diconnecting the mixnet client needs to be unified down this code path
+                // and merged with the mix tunnel one.
+                mixnet_client.disconnect().await;
+                return Err(err);
+            }
+        };
+
+        let bandwidth_controller =
+            BandwidthController::new(mixnet_client.clone(), self.task_manager.subscribe());
```

Comment:
Side comment for the future:
Imho we should switch BandwidthController to use CancellationToken. Also, BandwidthController calls `mixnet_client.disconnect()` I just spotted, not sure what the implication of that is ...

Reply:
Yeah what BandwidthController does is it seems that it just waits for task manager to shutdown before killing the mixnet client. I am not quite sure why it's done this way, this could be re-arranged.